lib/iris/common/metadata.py

```diff
@@ -54,6 +57,55 @@
 logger = get_logger(__name__, fmt="[%(cls)s.%(funcName)s]")
 
 
+_num2date_original = cf_units.Unit.num2date
+
+
+def _num2date_to_nearest_second(
+    self,
+    time_value,
+    only_use_cftime_datetimes=True,
+    only_use_python_datetimes=False,
+):
+    # Used to monkey-patch the cf_units.Unit.num2date method to round to the
+    #  nearest second, which was the legacy behaviour. This is under a FUTURE
+    #  flag - users will need to adapt to microsecond precision eventually,
+    #  which may involve floating point issues.
+    def _round(date):
+        if date.microsecond == 0:
+            return date
+        elif date.microsecond < 500000:
+            return date - timedelta(microseconds=date.microsecond)
+        else:
+            return (
+                date + timedelta(seconds=1) - timedelta(microseconds=date.microsecond)
+            )
+
+    result = _num2date_original(
+        self, time_value, only_use_cftime_datetimes, only_use_python_datetimes
+    )
+    if FUTURE.date_microseconds is False:
+        message = (
+            "You are using legacy date precision for Iris units - max "
+            "precision is seconds. In future, Iris will use microsecond "
+            "precision - from cf-units version 3.3 - which may affect core "
+            "behaviour. To opt-in to the "
+            "new behaviour, set `iris.FUTURE.date_microseconds = True`."
+        )
+        warnings.warn(message, category=FutureWarning)
+
+        if hasattr(result, "shape"):
+            vfunc = np.vectorize(_round)
+            result = vfunc(result)
+        else:
+            result = _round(result)
+
+    return result
+
+
+# See the note in _num2date_to_nearest_second.
+cf_units.Unit.num2date = _num2date_to_nearest_second
```

Comment:
This will change the function even when used outside of Iris, which may be unexpected to users. Maybe it would be cleaner to only change it in those places where it is used in iris internally.

Reply:
I think for this to work, units in iris would have to be a subclass of cf_units.Unit . This does seem like a cleaner way of doing things.

Reply:
If everything works out nicely, it might just be a case of modifying the class and instantiating it here: https://github.com/SciTools/iris/blob/01bbdf68524a9c6375b3f7b1ce52ec7385d404a3/lib/iris/common/mixin.py#L209-L210

Reply:
5c8fb3f9947a02999799efec19f287bb69c7e8d1

My tests still work, and the docstrings and hints still come from the original `Unit` class thankfully.